recon/src/recon.rs

```diff
@@ -21,114 +25,141 @@ use tracing::{instrument, trace};
-            let len = self.store.len();
-            match len {
-                0 => {
-                    // this should be unreachable, but we explicitly add this case to be clear.
-                }
-                1 => {
-                    // -> (only_key)
-                    response.keys.push(first.to_owned());
-                }
-                2 => {
-                    // -> (first 0 last)
-                    response.keys.push(first.to_owned());
-                    response.hashes.push(H::identity());
-                    response.keys.push(last.to_owned());
-                }
-                _ => {
-                    // -> (first h(middle) last)
-                    response.keys.push(first.to_owned());
-                    response.hashes.push(self.store.hash_range(first, last));
-                    response.keys.push(last.to_owned());
+    pub fn initial_messages(&self) -> Result<Vec<Message<K, H>>> {
+        let interests = self.interests()?;
+        let mut messages = Vec::with_capacity(interests.len());
+        for range in interests {
+            let mut response = Message::new(&range.start, &range.end);
+            if let Some((first, last)) = self.store.first_and_last(&range.start, &range.end) {
+                // Cannot use store.len, we need the count between first and last
+                let len = self.store.len();
+                match len {
+                    0 => {
+                        // this should be unreachable, but we explicitly add this case to be clear.
+                    }
+                    1 => {
+                        // -> (only_key)
+                        response.keys.push(first.to_owned());
+                    }
+                    2 => {
+                        // -> (first 0 last)
+                        response.keys.push(first.to_owned());
+                        response.hashes.push(H::identity());
+                        response.keys.push(last.to_owned());
+                    }
+                    _ => {
+                        // -> (first h(middle) last)
+                        response.keys.push(first.to_owned());
+                        response.hashes.push(self.store.hash_range(first, last));
+                        response.keys.push(last.to_owned());
+                    }
                 }
             }
+            messages.push(response)
         }
-        response
+        Ok(messages)
     }
 
     /// Process an incoming message and respond with a message reply.
-    pub fn process_message(&mut self, received: &Message<K, H>) -> Result<Response<K, H>> {
+    #[instrument(skip_all, ret)]
+    pub fn process_messages(&mut self, received: &[Message<K, H>]) -> Result<Response<K, H>> {
+        // First we must find the intersection of interests.
+        // Then reply with a message per intersection.
+        let mut intersections: Vec<(RangeOpen<K>, BoundedMessage<K, H>)> = Vec::new();
+        for range in self.interests.interests()? {
```

Comment:
Can we put a TODO here to explore more efficient alternatives to this implementation, with a ticket?

Reply:
Sure, I have some rough thoughts about using  a hierarchical tree something like (https://en.wikipedia.org/wiki/Bounding_volume_hierarchy) but in a single dimension. I'll add these to the TODO